proposals/ytt/005-semver/README.md

```diff
@@ -0,0 +1,217 @@
+error in the case of a malformed semver. If the operation succeeds, it returns a `struct` representing the semver.
+
+Now, it is possible to read the _major_, _minor_, _patch_ versions as integers, as well as the _pre-release_
+version and _build metadata_ as a strings. This enables authors to change its representation at will.
+
+It is also possible to programmatically construct a semver `struct` by providing _major_, _minor_, _patch_ versions as
+required integers, as well as the _pre-release_ version and _build metadata_ as optional strings. This enables authors
+to construct semvers either from values known in advance or which are received in a representation other than a semver.
+
+Given two semver `structs`, they can be compared. This enables authors to test if a semver is large or small enough,
+or whether it is within a range.
+
+For its implementation, the suggestion is to use [blang/semver](https://github.com/blang/semver). It is a
+well-established module, contains an API to address all conceivable case and is already used by Carvel's `vendir` for
+its `tools sort-semver` subcommand.
+
+### Goals and Non-goals
+
+* Enable configuration authors to programmatically work with semvers out-of-the-box.
+
+* Education around and proliferation of the semver scheme is not considered to be `ytt`'s responsibility.
+
+### Specification / Use Cases
+
+#### semver·version
+
+`semver.version([major=int, minor=int, patch=int, prerelease=string, metadata=string])` returns a struct representing a
+semantic version. The fields `major`, `minor` and `patch` are strings and default to `0`. The fields `prerelease` and
+`metadata` are strings and default to `None`. If any of the arguments are of the wrong type or contain forbidden
+characters, it is a dynamic error.
+
+```yaml
+#@ load("@ytt:semver", "semver")
+
+#@ semver.version()                                        #! 0.0.0
+#@ semver.version(1, 2, 3)                                 #! 1.2.3
+#@ semver.version(1, 2, 3, "beta")                         #! 1.2.3-beta
+#@ semver.version(1, 2, 3, metadata="build.5")             #! 1.2.3+build.5
+#@ semver.version(1, 2, 3, "next", "nightly")              #! 1.2.3-next+nightly
+#@ semver.version(minor=1, prerelease="alpha")             #! 0.1.0-alpha
+#@ semver.version(1, 2, -5)                                #! ⚡️ error
+#@ semver.version(prerelease="$send-ytt-coins$")           #! ⚡️ error
+#@ semver.version(meta="__dunder-score")                   #! ⚡️ error
+```
+
+Use cases:
+
+```yaml
+#@ load("@ytt:semver", "semver")
+#@ load("@ytt:yaml", "yaml")
+#@ load("@ytt:json", "json")
+#@ load("@ytt:assert", "assert")
+
+#@ v = semver.version(1, 2, 3)
+#@ d = dict(major=v.major, minor=v.minor, patch=v.patch, prerelease=v.prerelease, metadata=v.metadata) 
+
+--- #@ d
+
+---
+version.yaml: #@ yaml.encode(d)
+version.json: #@ json.encode(d)
+
+#@ d.update(major=d["major"]+1)
+#@ next_major = semver.version(**d)
```

Comment:
Interesting to propose keeping modifications outside of the semver API, itself. It certainly keeps things simpler language-wise. But I also wonder how frequently version modification (will) occur in templated configuration. 🤔

Reply:
I held back on modifications, because the spec only defines bumping major, minor and patch versions. What would be the next prerelease? Some tools like [semantic release](https://github.com/semantic-release/semantic-release) have opinions.

However, it would still be interesting to explore modifications in the API. It would help the discussion too I think, when there's something concrete to talk about. I will amend the proposal.